Fix InMemoryStreamLog

xnull opened the pull request:
## Overview

Description:
Before #3112 InMemoryStreamLog didn't need any information about the disk, and #3112 tried to read disk info during creation, which caused errors in private infrastructure. I have changed FileSystemAgent in an exact way how it worked before #3112 changes. 

I have added:
Complete integration of "memory mode" into FileSystemAgent.
This fix takes into account that StreamLog can work in "in-memory" mode. 

Why should this be merged: 

Related issue(s) (if applicable): #<number>


## Checklist (Definition of Done):

- [x] There are no TODOs left in the code
- [x] [Coding conventions](https://github.com/CorfuDB/CorfuDB/wiki/Corfu-Style-Guidelines) (e.g. for logging, unit tests) have been followed
- [x] Change is covered by automated tests
- [x] Public API has Javadoc

xnull commented on infrastructure/src/main/java/org/corfudb/infrastructure/log/FileSystemAgent.java:
```diff
@@ -43,13 +45,23 @@
+        logSizeQuota = new ResourceQuota("LogSizeQuota", logSizeLimit);
         logSizeQuota.consume(initialLogSize);
+
         log.info("FileSystemAgent: {} size is {} bytes, limit {}", config.logDir, initialLogSize, logSizeLimit);
```
Yes we print this log line for both in-mem and disk modes

Jas0nch commented:
It seems like the change is to add a branch that the FileSystemAgent can run in memeory mode, which can prevent the error when doing verifications. But my concern is that is there any possibility that some tests from nsx require LogUnitServer to run in disk mode. Then it'll fall into the old path and cause the exception again.

Maithem commented:
> It seems like the change is to add a branch that the FileSystemAgent can run in memeory mode, which can prevent the error when doing verifications. But my concern is that is there any possibility that some tests from nsx require LogUnitServer to run in disk mode. Then it'll fall into the old path and cause the exception again.

Yes, I think there some tests that restart the node. In memory mode is not supported for clustering. If you have three nodes and restart a bunch of them - which i believe the tests do - then the system will get bricked.

xnull commented:
> It seems like the change is to add a branch that the FileSystemAgent can run in memory mode, which can prevent the error when doing verifications. But my concern is that is there any possibility that some tests from nsx require LogUnitServer to run in disk mode. Then it'll fall into the old path and cause the exception again.

This change will not affect those test, everything will work as before. I added FileSysetmAgent and this PR just adds in-mem implementation of fsagent according to what we have all the time. Now FsAgent work according to rules we already have in corfu. So, essentially I just added missing parts in #3121 pr

xnull commented:
> > It seems like the change is to add a branch that the FileSystemAgent can run in memeory mode, which can prevent the error when doing verifications. But my concern is that is there any possibility that some tests from nsx require LogUnitServer to run in disk mode. Then it'll fall into the old path and cause the exception again.
> 
> Yes, I think there some tests that restart the node. In memory mode is not supported for clustering. If you have three nodes and restart a bunch of them - which i believe the tests do - then the system will get bricked.

Maithem, you can see that I didn't change StreamLogFiles or anything else, there are changes only in InMemoryStreamLog wich makes it similar to as it was before. Nothing else has changed, there will not be any side effects. At least not because of this change

xnull commented:
> > It seems like the change is to add a branch that the FileSystemAgent can run in memeory mode, which can prevent the error when doing verifications. But my concern is that is there any possibility that some tests from nsx require LogUnitServer to run in disk mode. Then it'll fall into the old path and cause the exception again.
> 
> Yes, I think there some tests that restart the node. In memory mode is not supported for clustering. If you have three nodes and restart a bunch of them - which i believe the tests do - then the system will get bricked.

Maithem, if your suggestion is right, as I know all our IT tests are run in in-mem mode, they shouldn't work then?

My change doesn't bring any new functionality and doesn't change the way how tests run in corfu and nsx. If there are particular tests in nsx and they worked for many years already with or without in-mem mode, then they will work with my change.
